xmppserver/src/main/java/org/jivesoftware/openfire/session/LocalSession.java

```diff
@@ -198,14 +219,20 @@ public void setAddress(JID address){
      * @return The connection for this session
      */
     public Connection getConnection() {
-        if (conn == null) {
-            try {
-                conn.isClosed(); // This generates an NPE deliberately.
-            } catch (NullPointerException e) {
-                Log.error("Attempt to read connection of detached session: ", e);
+        lock.readLock().lock();
+        try {
+            if (conn == null)
+            {
+                try {
```

Comment:
Not a comment on your changes, but as you're here ...

replace this try/catch with a 

Log.error("Attempt to read connection of detached session", new IllegalStateException());

Reply:
done